crates/agent/src/reconcile/storage.rs

```diff
@@ -0,0 +1,517 @@
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use snops_checkpoint::CheckpointManager;
+use snops_common::{
+    api::AgentEnvInfo,
+    binaries::{BinaryEntry, BinarySource},
+    constant::{
+        LEDGER_BASE_DIR, LEDGER_PERSIST_DIR, NODE_DATA_DIR, SNARKOS_FILE, SNARKOS_GENESIS_FILE,
+        VERSION_FILE,
+    },
+    rpc::error::ReconcileError,
+    state::{HeightRequest, InternedId, ReconcileCondition, ReconcileStatus, TransferId},
+};
+use tokio::{process::Command, sync::Mutex, task::AbortHandle};
+use tracing::{error, info, trace};
+use url::Url;
+
+use super::{default_binary, get_genesis_route, DirectoryReconciler, FileReconciler, Reconcile};
+use crate::state::GlobalState;
+
+/// Download a specific binary file needed to run the node
+pub struct BinaryReconciler<'a> {
+    pub state: Arc<GlobalState>,
+    pub env_info: Arc<AgentEnvInfo>,
+    pub node_binary: Option<InternedId>,
+    /// Metadata about an active binary transfer
+    pub transfer: &'a mut Option<(TransferId, BinaryEntry)>,
+    /// Time the binary was marked as OK
+    pub ok_at: &'a mut Option<Instant>,
```

Comment:
The preference for lifetimes on Options is to never have a reference to an Option, but rather the insides of an Option.

Reply:
These are mutable references to options because it's mutating data from the parent state

Reply:
Are you replacing the option from Some to None? Otherwise the inside being mutable would the way to go.

Reply:
yes